File: 6/Assembler/SymbolTable.py
class SymbolTable:
    def __init__(self):
        self.symbol_table = {
            "SP": 0,
            "LCL": 1,
            "ARG": 2,
            "THIS": 3,
            "THAT": 4,
            "SCREEN": 16384,
            "KBD": 24576,
        }
        
        for i in range(16):
            self.symbol_table[f"R{i}"] = i

    def addEntry(self, symbol, address):
        self.symbol_table[symbol] = address

    def contains(self, symbol):
        return symbol in self.symbol_table

    def getAddress(self, symbol):
        return self.symbol_table[symbol]
    
    def __str__(self):
        result = ""
        for symbol, address in self.symbol_table.items():
            result += f"{symbol}: {address}\n"
            
        return result
    
if __name__ == "__main__":
    s = SymbolTable()
    print(s)

File: projects/6/HackAssembler/symboltable.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "symboltable.h"
#include "config.h"

static SymbolTable symbolTable[TABLE_MAX];
static int symbolCount = 0;


void initSymbolTable(void)
{
    symbolCount = 0;

    char buf[BUF_MAX];

    addSymbol("SP", 0);
    addSymbol("LCL", 1);
    addSymbol("ARG", 2);
    addSymbol("THIS", 3);
    addSymbol("THAT", 4);

    for (int i = 0; i < 16; i++) {
        snprintf(buf, BUF_MAX, "R%d", i);
        addSymbol(buf, i);
    }

    addSymbol("SCREEN", 16384);
    addSymbol("KBD", 24576);
}

void addSymbol(const char *symbol, int address)
{
    if (symbolCount >= TABLE_MAX) {
        fprintf(stderr, "Symbol table is full.\n");
        exit(1);
    }

    strncpy(symbolTable[symbolCount].symbol, symbol,SYMBOL_MAX);
    symbolTable[symbolCount].symbol[SYMBOL_MAX-1] = '\0';
    symbolTable[symbolCount].address = address;
    symbolCount++;
}

int containsSymbol(const char *symbol)
{
    for (int i = 0; i < symbolCount; i++) {
        if (!strcmp(symbolTable[i].symbol, symbol))
            return 1;
    }

    return 0;
}

int getAddress(const char *symbol)
{
    for (int i = 0; i < symbolCount; i++) {
        if (!strcmp(symbolTable[i].symbol, symbol))
            return symbolTable[i].address;
    }

    fprintf(stderr, "Undefined symbol: %s\n", symbol);
    exit(1);
}
